source/common/network/utility.cc

```diff
@@ -619,16 +612,27 @@ Api::IoCallUint64Result Utility::readFromSocket(IoHandle& handle,
   }
 
   if (handle.supportsMmsg()) {
-    const uint32_t num_packets_per_mmsg_call = 16u;
-    const uint32_t num_slices_per_packet = 1u;
-    absl::FixedArray<Buffer::InstancePtr> buffers(num_packets_per_mmsg_call);
+    const auto max_packet_size = udp_packet_processor.maxPacketSize();
+
+    // Buffer::ReservationSingleSlice is always passed by value, and can only be constructed
+    // by Buffer::Instance::reserve(), so this is needed to keep a fixed array
+    // in which all elements are legally constructed.
+    struct BufferAndReservation {
+      BufferAndReservation(uint64_t max_packet_size)
+          : buffer_(std::make_unique<Buffer::OwnedImpl>()),
+            reservation_(buffer_->reserveSingleSlice(max_packet_size, true)) {}
+
+      Buffer::InstancePtr buffer_;
+      Buffer::ReservationSingleSlice reservation_;
+    };
+    constexpr uint32_t num_packets_per_mmsg_call = 16u;
+    constexpr uint32_t num_slices_per_packet = 1u;
+    absl::InlinedVector<BufferAndReservation, num_packets_per_mmsg_call> buffers;
     RawSliceArrays slices(num_packets_per_mmsg_call,
                           absl::FixedArray<Buffer::RawSlice>(num_slices_per_packet));
-    for (uint32_t i = 0; i < num_packets_per_mmsg_call; ++i) {
-      buffers[i] = std::make_unique<Buffer::OwnedImpl>();
-      const uint64_t num_slices = buffers[i]->reserve(udp_packet_processor.maxPacketSize(),
-                                                      slices[i].data(), num_slices_per_packet);
-      ASSERT(num_slices == num_slices_per_packet);
+    for (uint32_t i = 0; i < num_packets_per_mmsg_call; i++) {
+      buffers.push_back(max_packet_size);
```

Comment:
Not sure how much it matters, but I think that before we used to allocate 4kb per 1500 UDP packet we attempt to read, now we allocate 16kb per UDP packet.